sorted_db/db.go

```diff
@@ -143,46 +160,97 @@ func makeCopy(s []byte) []byte {
 	return d
 }
 
-// Search uses a binary search looking for needle, and returns the full match line.
-// the needle should already have the record separator appended
-func (db *DB) Search(needle []byte) []byte {
-	db.RLock()
-
+// findFirstMatch performs a binary search to find the first record
+// that matches needle using the given isMatch function, or -1 if
+// no match is found.
+func (db *DB) findFirstMatch(needle []byte, isMatch func([]byte) bool) int {
 	needleLen := len(needle)
 
-	if db.size <= 0 {
-		panic("DB not Mapped")
-	}
-	// binary search to find the index that matches our needle (starting at the previous line)
-	// note: this could be more efficient if we wrote our own search as we could skip data we've checked
-	// isntead of checking potentially more indexes here. Because page sizes is 4k this should hopefully
-	// matter less
-	i := sort.Search(db.size, func(i int) bool {
+	// binary search to find the index that matches our needle,
+	// starting at the previous line.
+	// note: this could be more efficient if we wrote our own search as we could
+	// skip data we've checked instead of checking potentially more indexes here.
+	// Because page size is 4k this should hopefully matter less.
+	return sort.Search(db.size, func(i int) bool {
 		// find previous line starting point
 		atomic.AddUint64(&db.seekCount, 1)
-		previous := lastIndexByte(db.data, i, db.LineEnding)
-		if previous == -1 {
-			previous = 0
-		} else {
-			previous++ // eat the line ending
-		}
+
+		previous := db.beginningOfLine(i)
+
 		// make sure we have space before end of the buffer
 		if previous+1+needleLen > db.size {
 			return false
 		}
-		return bytes.Compare(db.data[previous:previous+needleLen], needle) >= 0
+		endOfKey := indexByte(db.data, previous, db.size, db.RecordSeparator)
+		if endOfKey < 0 {
```

Comment:
There is actually a bug here (if the next record separator comes after the line ending) so I'm rewriting this chunk.